docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,111 @@
+
+Here is the proposed set of `ConnectionManager` states and how they map to the `Controller` states. Please note that some of these states may go away once we are further along the implementation (e.g. we may not need `ConnectionStateSilentSwitching` anymore):
+
+```c++
+  enum VPNClientState {
+    VPNClientStateOff,          // VPN is off
+    VPNClientStateConnecting,   // VPN is perform connection and 
+                                // subscription probes in ConnectionState
+    VPNClientStateOn,           // VPN is on
+  }
+
+  enum ConnectionState {
+    ConnectionStateOff,                // Corresponds to client state:
+                                       // VPNClientStateOff
+    ConnectionStateInternetProbe,      // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateServerProbe,        // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateFirewall,           // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateCaptivePortal,      // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateCheckSubscription,  // Corresponds to client state:
+                                       // VPNClientStateConnecting
+    ConnectionStateOn,                 // Corresponds to client state:
+                                       // VPNClientStateOn
+  };
+```
+
+If at any point any of the checks in `ConnectionManager::ConnectionStates` fails then we will emit a signal to the frontend explaining what exactly the issue is. For example, if we are in `ConnectionManager::ConnectionStateInternetProbe` and the internet probe fails, we will notify the frontend via `emit connectionStateInternetProbeFailed()` signal. This allows the UI to notify the user that the VPN activation failed due to their network connectivity issue and may offer troubleshooting advice such as checking that their wifi/cellular is enabled. It is worth noting that in any given circumstance, we will not deactivate the VPN on the backend regardless of the failure. 
+
+## Metrics and Measurements
+
+This change also means that prior to the activation of the VPN, there are multiple checks and probes that need to succeed; this will take some time (hopefully no more than a couple of seconds), and as long as we communicate to the user via front end prompts about what is happening, this is acceptable. 
+
+To confidently determine that these checks do not delay the VPN activation by an unacceptable amount, we need to measure the duration of the _time elapsed from toggling on the VPN to activation_. This is the only relevant metric, given that in any other scenario the checks run in the background and would not affect the user experience.
```

Comment:
Its worth noting that telemetry is not only useful for measuring metrics related to the user experience. It can be very useful for technical data as well to measure the health of this system. I am not saying we need telemetry for that here, but we should at least reword this sentence and give it a second thought if there are metrics we can add to this new system that would help us asses it's health after release.

Reply:
>To confidently determine that these checks do not delay the VPN activation by an unacceptable amount, we need to measure the duration

That's what this bit is referring to, what else in the context of system health would you wan to see in this section?

Reply:
> This is the only relevant metric, given that in any other scenario the checks run in the background and would not affect the user experience.

The proposed telemetry should help us understand system health. But this sentence is implying that unless something affects user experience we do not need to monitor it with telemetry and this is not correct. I am calling out that if the telemetry design was thought out with this in mind it should be revisited without this assumption. I did not have any extra telemetry in mind when I said it, it was more of a conceptual comment.